components/raftstore-v2/src/operation/command/admin/conf_change.rs

```diff
@@ -0,0 +1,509 @@
+        // TODO: use local histogram metrics
+        PEER_PROPOSE_LOG_SIZE_HISTOGRAM.observe(data_size as f64);
+        info!(
+            self.logger,
+            "propose conf change peer";
+            "changes" => ?changes.as_ref(),
+            "kind" => ?ConfChangeKind::confchange_kind(changes.as_ref().len()),
+        );
+
+        let last_index = self.raft_group().raft.raft_log.last_index();
+        self.raft_group_mut()
+            .propose_conf_change(ProposalContext::SYNC_LOG.to_vec(), cc)?;
+        let proposal_index = self.raft_group().raft.raft_log.last_index();
+        if proposal_index == last_index {
+            // The message is dropped silently, this usually due to leader absence
+            // or transferring leader. Both cases can be considered as NotLeader error.
+            return Err(Error::NotLeader(self.region_id(), None));
+        }
+
+        Ok(proposal_index)
+    }
+
+    pub fn on_apply_res_conf_change(&mut self, conf_change: ConfChange) {
+        // TODO: cancel generating snapshot.
+
+        // Snapshot is applied in memory without waiting for all entries being
+        // applied. So it's possible conf_change.index < first_index.
+        if conf_change.index >= self.raft_group().raft.raft_log.first_index() {
+            match self.raft_group_mut().apply_conf_change(&conf_change.conf_change) {
+                Ok(_)
+                // PD could dispatch redundant conf changes.
+                | Err(raft::Error::NotExists { .. }) | Err(raft::Error::Exists { .. }) => (),
+                _ => unreachable!(),
+            }
+        }
+
+        let remove_self = conf_change.region_state.get_state() == PeerState::Tombstone;
+        self.storage_mut()
+            .set_region_state(conf_change.region_state);
+        if self.is_leader() {
+            info!(
+                self.logger,
+                "notify pd with change peer region";
+                "region" => ?self.region(),
+            );
+            let demote_self = tikv_util::store::is_learner(self.peer());
+            if remove_self || demote_self {
+                warn!(self.logger, "removing or demoting leader"; "remove" => remove_self, "demote" => demote_self);
+                let term = self.term();
+                self.raft_group_mut()
+                    .raft
+                    .become_follower(term, raft::INVALID_ID);
+            } else if conf_change.changes.iter().any(|c| {
+                matches!(
+                    c.get_change_type(),
+                    ConfChangeType::AddNode | ConfChangeType::AddLearnerNode
+                )
+            }) {
+                // Speed up snapshot instead of waiting another heartbeat.
+                self.raft_group_mut().ping();
+                self.set_has_ready();
+            }
+        }
+        if remove_self {
+            self.mark_for_destroy(None);
+        }
+    }
+}
+
+impl<EK: KvEngine, R> Apply<EK, R> {
+    #[inline]
+    pub fn apply_conf_change(
```

Comment:
let's remove the support of `ChangePeerv1` in raftstore v2

Reply:
I'm afraid it's impossible if we need to support upgrading from old version.

Reply:
> I'm afraid it's impossible if we need to support upgrading from old version.

What's the default version in 6.1 TLS? We can require the minimal version to upgrade from is 6.1. If it can help remove v1 support.

Reply:
Both v1 and v2 are used.

Reply:
Sadly, we only enforce all operators go through v2 in v6.3 https://github.com/tikv/pd/pull/5408